packages/@aws-cdk/aws-servicecatalog/lib/private/association-manager.ts

```diff
@@ -0,0 +1,90 @@
+ */
+type AssociationOrConstraint = CfnPortfolioProductAssociation | CfnResourceUpdateConstraint;
+
+export class AssociationManager {
+  public static associateProductWithPortfolio(scope: cdk.Resource, portfolio: IPortfolio, product: IProduct): void {
+    const associationKey = this.getAssociationKey(scope, portfolio, product);
+    if (this.checkConstraintOrAssocationExists(associationKey, Constraints.ASSOCIATION)) {
+      const association = new CfnPortfolioProductAssociation(scope, `PortfolioProductAssociation${associationKey}`, {
+        portfolioId: portfolio.portfolioId,
+        productId: product.productId,
+      });
+
+      const associationValue = new Map<Constraints, AssociationOrConstraint>();
+      associationValue.set(Constraints.ASSOCIATION, association);
+      this.associationMap.set(associationKey, associationValue);
+    }
+  }
+
+  public static allowTagUpdates(
+    scope: cdk.Resource, portfolio: IPortfolio, product: IProduct,
+    options: TagUpdateConstraintOptions,
+  ): void {
+    InputValidator.validateLength(this.prettyPrintAssociation(portfolio, product), 'description', 0, 2000, options.description);
+    const associationKey = this.ensureProductIsAssociatedWithPortfolio(scope, portfolio, product);
+    if (this.checkConstraintOrAssocationExists(associationKey, Constraints.RESOURCE_UPDATE)) {
+      const constraint = new CfnResourceUpdateConstraint(scope, `ResourceUpdateConstraint${associationKey}`, {
+        acceptLanguage: options.messageLanguage,
+        description: options.description,
+        portfolioId: portfolio.portfolioId,
+        productId: product.productId,
+        tagUpdateOnProvisionedProduct: options.allowUpdatingProvisionedProductTags === false ? 'NOT_ALLOWED' : 'ALLOWED',
+      });
+
+      // Add dependsOn to force proper order in deployment.
+      constraint.addDependsOn(this.associationMap.get(associationKey)!.get(Constraints.ASSOCIATION)!);
+      this.associationMap.get(associationKey)!.set(Constraints.RESOURCE_UPDATE, constraint);
+    } else {
+      throw new Error(`Cannot have multiple resource update constraints for association ${this.prettyPrintAssociation(portfolio, product)}`);
+    }
+  }
+
+  private static associationMap = new Map<string, Map<Constraints, AssociationOrConstraint>>();
+
+  private static getAssociationKey(scope: cdk.Resource, portfolio: IPortfolio, product: IProduct): string {
+    return hashValues(portfolio.node.addr, product.node.addr, scope.stack.node.addr);
+  }
+
+  private static ensureProductIsAssociatedWithPortfolio(scope: cdk.Resource, portfolio: IPortfolio, product: IProduct): string {
+    const associationKey = this.getAssociationKey(scope, portfolio, product);
+    // We can still call this because it will check if the association already exists before creating a new one
+    this.associateProductWithPortfolio(scope, portfolio, product);
+    return associationKey;
+  }
```

Comment:
How about you just change `associateProductWithPortfolio()` to return the association key, and the `Map<Constraint, AssociationOrConstraint>`?

```ts
  public static associateProductWithPortfolio(
     scope: cdk.Resource, portfolio: IPortfolio, product: IProduct,
  ): { associationKey: string, constraints: Map<Constraint, AssociationOrConstraint> } {
```

You would be able to get rid of this method, and also would get rid of the problem of repeating the `this.associationMap.get(associationKey)!` expression above.

(Feel free to introduce an interface for the return type if that's what you'd prefer)

Reply:
With using the node child I can actually just return the association key string and remove the method.  I like this.